Automattic/jetpack
# Add a generator for js-packages

~~*WARNING:* this changes file merge logic for the generator. This means that by default when you generate over an existing package, files with the same names will be overwritten. Since we are using source control I figured this would be the better behaviour, so you can use `git diff` to pick changes.~~

~~I made this change so that we could have js-package skeleton files that override common skeleton files, for example to add the `test-js` composer script.~~ 


----

This was extracted out of the [Standalone My Plans Page PR](https://github.com/Automattic/jetpack/pull/20543).

It adds a js-packages generator which includes basic test-js support.

~~It also adds a feature to the generator where matching files in package-specific skeleton directories automatically override common files. This is what allows the generator to exclude test-php and include test-js instead, for example.~~

#### Changes proposed in this Pull Request:
* Add a `jetpack generate` command for js-packages

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
* `jetpack generate`
* choose `js-packages`
* choose a name like `testpackage`
* choose a description like `testpackage test`
* accept the default build steps (we should adjust these for js-packages later)
* accept the default wordbless question (we should remove this question for js-packages later)
* say "n" to repo mirror
* run `git status` to confirm you have a new uncommitted dir called `projects/js-packages/testpackage/`
* `cd projects/js-packages/testpackage/`
* `composer run test-js`

After the last command, the packages should be installed and it should run 0 tests, since there are no test files yet.

You should see these lines in composer.json:

```
  "name": "automattic/jetpack-testpackage",
  "description": "testpackage test",
```

as well as the added development/production build scripts:

```
    "build-production": "echo 'Add your build step to composer.json, please!'",
    "build-development": "echo 'Add your build step to composer.json, please!'"
```

Similarly you should see these lines near the top of package.json:

```
  "name": "@automattic/jetpack-testpackage",
  "version": "0.1.0-alpha",
  "description": "testpackage test",
```

and a test script like this one:

```
"scripts": {
    "test": "NODE_ENV=test NODE_PATH=tests:. js-test-runner --jsdom --initfile=test-main.jsx 'glob:./!(node_modules)/**/test/*.@(jsx|js)'"
  },
```
